File: typescript/src/parser.ts
import {ElevatorCall} from "./elevator-call";

export class Parser {

    parseSingleCall(callToken: string): ElevatorCall {
        let floors = callToken.split("-");
        let numbers = floors.map(x => this.mapFloor(x));
        let origin: number = numbers[0];
        let destination: number = numbers[1];
        return new ElevatorCall(origin, destination)
    }

    private mapFloor(x: string) {
        switch (x) {
            case "G":
            case "g":
                return 0
            case "B":
            case "b":
                return -1
            default :
                return parseInt(x, 10);
        }
    }

    public parseCalls(calls: string): ElevatorCall[] {
        try {
            const callTokens = calls.split(" ");
            return callTokens.map(x => this.parseSingleCall(x))
        } catch (error) {
            return [];
        }
    }
}

File: python/elevator_kata/parser.py
from elevator_kata.elevator import ElevatorCall


def __map_floor(floor: str) -> int:
    match floor:
        case "G" | "g":
            return 0
        case "B" | "b":
            return -1
        case _:
            return int(floor)


def __parse_call(unparsed_call: str) -> ElevatorCall:
    origin, destination = unparsed_call.split("-")
    origin_floor = __map_floor(origin)
    destination_floor = __map_floor(destination)
    return ElevatorCall(origin_floor, destination_floor)


def parse_calls(calls: str) -> list[ElevatorCall]:
    try:
        call_tokens = calls.split(" ")
        return [__parse_call(unparsed_call) for unparsed_call in call_tokens]
    except ValueError:
        return []
